### PR TITLE
first version of training loop for dyna

### PR DESCRIPTION
**Pull request by Misterhoonster**
Could you take a look at my starter training loop logic?

### CODE REVIEW

File: examples/crafter/dynax.py [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+
+# --- Dataclasses ---
+
+# for single env, single timestep
+@struct.dataclass
+class TimeStep:
+    observation: jax.Array
+    reward: jax.Array
+    done: jax.Array
+    env_state: struct.PyTreeNode # Raw environment state
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:23]: call this state to be compatible with other libraries

File: examples/crafter/dynax.py [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+
+@struct.dataclass
+class AgentState:
+    # Represents the recurrent state of the agent (e.g., LSTM state)
+    rnn_state: struct.PyTreeNode # Could be a tuple (h, c) or just h
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:28]: can remove this if you want. but not bad. future work would probably do this

File: examples/crafter/dynax.py [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+    action: jax.Array       # Action taken at time t (a_t)
+    agent_state: AgentState # Agent's RNN state *before* processing timestep t (h_{t-1})
+
+class CustomTrainState(TrainState):
+    target_network_params: flax.core.FrozenDict
+    timesteps: int
+    n_updates: int
+
+@struct.dataclass
+class RunnerState():
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:43]: I would not use this and just use tuples

File: examples/crafter/dynax.py [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+@struct.dataclass
+class WorldModelState:
+    # Placeholder for potential learned world model state
+    internal_state: Optional[jax.Array] = None
+
+@struct.dataclass
+class WorldModelOutput:
+    next_observation: jax.Array
+    reward: jax.Array
+    done: jax.Array # Predicted termination
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:64]: the convention in the RL community is to use discount factor. but the convention outside of RL is to use done. done = 1 - discount. 

both are fine.

File: examples/crafter/dynax.py [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+    # Potentially next env_state if model predicts it directly
+    next_env_state: Optional[struct.PyTreeNode] = None
+
+
+# --- Network Definition ---
+
+class RecurrentQNetwork(nn.Module):
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:71]: Have a single Network with
1. setup
2. initialize_carry
3. __call__ for when its used with read data
4. apply_model

File: examples/crafter/dynax.py [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+    @staticmethod
+    def initialize_carry(batch_size, hidden_size):
+        # Use a dummy key since the default state init fn is just zeros.
+        return nn.GRUCell.initialize_carry(
+            jax.random.PRNGKey(0), (batch_size,), hidden_size
+        )
+
+# --- World Model Definition ---
+
+class WorldModel(nn.Module):
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:132]: I would remove?

File: examples/crafter/dynax.py [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+        jax.random.split(rng, n_envs), env_params
+    )
+    vmap_step = lambda n_envs: lambda rng, env_state, action: jax.vmap(
+        env.step, in_axes=(0, 0, 0, None)
+    )(jax.random.split(rng, n_envs), env_state, action, env_params)
+
+    def train(rng):
+        # Initialize environment
+        rng, _rng = jax.random.split(rng, 2)
+        init_obs, env_state = vmap_reset(config["N_ENVS"])(_rng)
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:246]: Recommend using an environment that by default inputs and returns TimeStep objects. you can wrap the environment to do this.

Comment (wcarvalho) [reply examples/crafter/dynax.py:246]: https://github.com/wcarvalho/nicewebrl/blob/c6913ff82ecefb5c35074d0ec777a59ab1c46660/nicewebrl/nicejax.py#L227

File: examples/crafter/dynax.py [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+
+        # --- Miscellaneous ---
+        "SEED": 42,
+        "LEARNER_LOG_PERIOD": 100,  # How many LEARNER UPDATES between logging losses/metrics
+        "EVAL_LOG_PERIOD": 100,     # How many LEARNER UPDATES between running evaluation episodes
+        "GRADIENT_LOG_PERIOD": 500, # How many LEARNER UPDATES between logging gradients (0 to disable)
+    }
+
+    # Create env and env_params
+    env = make_craftax_env_from_name("Craftax-Symbolic-v1", auto_reset=True)
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:546]: can wrap here: https://github.com/wcarvalho/nicewebrl/blob/c6913ff82ecefb5c35074d0ec777a59ab1c46660/nicewebrl/nicejax.py#L227

Comment (wcarvalho) [reply examples/crafter/dynax.py:546]: turn autoreset=False

File: examples/crafter/dynax.py [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+            reward=jnp.array(0.0),
+            done=jnp.array(0),
+            env_state=None
+        )
+        dummy_agent_state = AgentState(rnn_state=agent.initialize_carry((1,)))
+        online_params = agent.init(init_rng, dummy_agent_state, dummy_timestep, rng)
+        target_params = online_params
+        
+        # Initialize GroundTruthWorldModel, get initial (empty) params
+        world_model = GroundTruthWorldModel(env=env, env_params=env_params)
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:266]: remove

File: examples/crafter/dynax.py [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+            q_network=agent,
+            world_model=world_model,
+            config=config,
+            simulation_policy_fn=simulation_policy
+        )
+        
+        # Initialize RunnerState
+        rng, env_rng = jax.random.split(rng)
+        initial_timestep = env.reset(env_rng, env_params)
+        initial_agent_state = AgentState(rnn_state=agent.initialize_carry((1,)))
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:336]: this is probably wrong. (1,) can't be batch_size or RNN size.

Comment (wcarvalho) [reply examples/crafter/dynax.py:336]: make sure compatible with 32 parallel envs

File: examples/crafter/dynax.py [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+            runner_state = RunnerState(
+                train_state=train_state,
+                env_timestep=timestep,
+                agent_state=agent_state,
+                buffer_state=buffer_state,
+                rng=rng
+            )
+
+            # --- 3. Learning Update ---
+            model_params = {} # Empty for GroundTruthWorldModel
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:378]: what's this?

File: examples/crafter/dynax.py [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+                rng=rng
+            )
+            
+            return runner_state, {}
+
+        # JIT the _train_step function
+        _train_step = jax.jit(_train_step)
+
+        # Initialize the initial runner_state
+        initial_runner_state = RunnerState(
```
Comment (wcarvalho) [inline examples/crafter/dynax.py:458]: remove runnerState

**Comment by Misterhoonster** [comment]
Hey Wilka, just updated my dyna file with your suggested changes!

**Comment by Misterhoonster** [comment]
Pushed the new version with the Loss class! Can you take a look, please?
